### PR TITLE
Replace gen.coroutine with async/await in core

### PR DESCRIPTION
**Pull request by mrocklin**
This replaces the use of Tornado `gen.coroutine`s and the `yield` keyword with `async def` functions and the `await` keyword.  Additionally, because these have slightly different semantics this also changes a few internal bits to smooth over this transition.

This is still a work in progress.  The current implementation introduces a couple of small bugs and the conversion is not complete.

cc @jcrist (who requested this a while ago) and @jacobtomlinson who might find this easier to work with than gen.coroutine.

### CODE REVIEW

**Comment by mrocklin** [comment]
This is getting close to mergeable, and includes fixes for issues in master.  I'm going to try to clean things up here today or tomorrow and hopefully merge in shortly afterwards.

**Comment by mrocklin** [comment]
There were some cases where I've decided to keep gen.coroutine semantics.  In particular asyncio complains if we leave a task un-awaited in a way that it is bothersome for users.  In a couple of places I didn't find a nice way to await every task (such as in the case of the TCP Comm.write method) so I left these in the tornado style.

**Comment by mrocklin** [comment]
Tests are now starting to look good :)

**Comment by mrocklin** [comment]
OK, I'm going to pronounce this as "good enough" and merge in.  I'll probably do a follow-up in a couple of days.

File: distributed/scheduler.py [hunk 1]
```diff
@@ -1305,7 +1314,6 @@ def close_worker(self, stream=None, worker=None, safe=None):
     # Stimuli #
     ###########
 
-    @gen.coroutine
```
Comment (lr4d) [inline distributed/scheduler.py]: `async` is not used here. Is this intentional?

Comment (mrocklin) [reply distributed/scheduler.py]: Probably?  Is this causing some concrete issue for you?  Does this function need to be asynchronous?

**Pull request merged**